crates/pathfinder/src/state/sync.rs

```diff
@@ -871,57 +890,135 @@ async fn download_verify_and_insert_missing_classes<
 
     // For each missing, download, verify and insert definition.
     for class_hash in missing {
-        let definition = sequencer
-            .class_by_hash(class_hash)
-            .await
-            .with_context(|| format!("Downloading class {}", class_hash.0))?;
-
-        // Calculate the class hash. This can be expensive, so perform in a blocking task.
-        let extract = tokio::task::spawn_blocking(move || -> anyhow::Result<_> {
-            let hash = compute_class_hash(&definition)?;
-            Ok((definition, hash))
-        });
-        let (definition, hash) = extract
-            .await
-            .context("Parse class definition and compute hash")??;
+        let class = download_class(&sequencer, class_hash).await?;
+
+        match class {
+            DownloadedClass::Cairo(class) => {
+                tokio::task::block_in_place(|| {
+                    let transaction =
+                        connection.transaction_with_behavior(TransactionBehavior::Immediate)?;
+                    ContractCodeTable::insert_compressed(&transaction, &class)?;
+                    transaction.commit()?;
+                    anyhow::Result::<()>::Ok(())
+                })
+                .with_context(|| format!("Insert class definition with hash: {:?}", class.hash))?;
+            }
+            DownloadedClass::Sierra(sierra, casm) => {
+                let compiled_class_hash = state_update
+                    .state_diff
+                    .declared_classes
+                    .iter()
+                    .find_map(|declared_class| {
+                        if declared_class.class_hash.0 == class_hash.0 {
+                            Some(declared_class.compiled_class_hash)
+                        } else {
+                            None
+                        }
+                    })
+                    .ok_or_else(|| anyhow::anyhow!("Sierra class hash not in declared classes"))?;
+                tokio::task::block_in_place(|| {
+                    let transaction =
+                        connection.transaction_with_behavior(TransactionBehavior::Immediate)?;
+                    ContractCodeTable::insert_compressed(&transaction, &sierra)?;
+                    CasmClassTable::upsert_compressed(&transaction, &casm, &compiled_class_hash)?;
```

Comment:
Just thinking out loud: IIRC calculating casm hash for verification does not make sense here as a mismatch is still possible as the latest compiler version may not be the one used for calculating the original casm hash when the class was declared :thinking: .

Reply:
Exactly. What CASM hash we calculate is irrelevant, because the class commitment tree contains the one the sequencer returns.

As discussed with Ariel the only place where we _should_ be verifying that the CASM hash matches is in `starknet_estimateFee` so that we can reject an invalid transaction with a meaningful error message.

Reply:
Can we add a comment to that effect?

Reply:
I've added a comment.